src/storage/src/hummock/shared_buffer/shared_buffer_batch.rs

```diff
@@ -364,40 +650,314 @@ impl SharedBufferBatch {
+                            } else {
+                                self.current_value_idx = idx;
+                            }
+                        }
+                        // Seek to one item before the seek partition_point:
+                        // If i == 0, the iterator will be invalidated with self.current_idx ==
+                        // self.inner.len().
+                        Err(i) => {
+                            self.current_key_idx = self.inner.len() - i;
+                            self.current_value_idx = self.current_values_len() - 1;
+                        }
+                    }
+                }
+            }
+            Ok(())
+        }
+    }
+
+    fn collect_local_statistic(&self, _stats: &mut crate::monitor::StoreLocalStatistic) {}
+}
+
+/// A snapshot iterator for a given batch(Imm)
+/// Only scan value with epoch <= `read_epoch`
+pub struct SharedBufferBatchSnapshotIterator<D: HummockIteratorDirection> {
+    inner: Arc<SharedBufferBatchInner>,
+    current_key_idx: usize,
+    table_id: TableId,
+    read_epoch: HummockEpoch,
+    _phantom: PhantomData<D>,
+}
+
+impl<D: HummockIteratorDirection> SharedBufferBatchSnapshotIterator<D> {
     pub(crate) fn new(
         inner: Arc<SharedBufferBatchInner>,
         table_id: TableId,
-        epoch: HummockEpoch,
+        read_epoch: HummockEpoch,
     ) -> Self {
         Self {
             inner,
-            current_idx: 0,
+            current_key_idx: 0,
             table_id,
-            epoch,
+            read_epoch,
             _phantom: Default::default(),
         }
     }
 
-    fn current_item(&self) -> &SharedBufferItem {
+    pub(crate) fn current_entry(&self) -> &SharedBufferVersionedEntry {
         assert!(self.is_valid());
         let idx = match D::direction() {
-            DirectionEnum::Forward => self.current_idx,
-            DirectionEnum::Backward => self.inner.len() - self.current_idx - 1,
+            DirectionEnum::Forward => self.current_key_idx,
+            DirectionEnum::Backward => self.inner.len() - self.current_key_idx - 1,
         };
         self.inner.get(idx).unwrap()
     }
+
+    pub fn epoch(&self) -> HummockEpoch {
+        self.read_epoch
+    }
+
+    fn check_visibility_and_seek(&mut self) {
+        // move to a key that has visible versions
+        loop {
+            if !self.is_valid() {
+                break;
+            }
+            let values = &self.current_entry().1;
+            debug_assert!(!values.is_empty());
+            let min_epoch = values.last().unwrap().0;
+            // read_epoch >= min_epoch there exists a visible version
+            if min_epoch <= self.read_epoch {
+                break;
+            }
+            self.current_key_idx += 1;
+        }
+    }
 }
 
-impl<D: HummockIteratorDirection> HummockIterator for SharedBufferBatchIterator<D> {
+impl<D: HummockIteratorDirection> HummockIterator for SharedBufferBatchSnapshotIterator<D> {
```

Comment:
I don't think we need a `SharedBufferBatchSnapshotIterator` for read epoch filtering since filtering invisible epoch is the responsibility of user key iterator. Although it is possible do epoch filtering push-down for all hummock iterators, we don't see the needs so we didn't do it (e.g. no epoch filter for SstableIterator, MergeIterator).